[AutoTVM] [TVMC] Visualize tuning progress

PhilippvK opened the pull request:
This change adds a callback for visualizing the FLOPS during tuning (AutoTVM & AutoScheduler (**NEW**)) over time using Matplotlib.

## Usage

### With Python

Create callback (`multi=False`: 1 window per Task, `multi=True`: all tasks in a single window):

```
visualize_callback = autotvm.callback.visualize_progress(task_idx, live=True, out_path="plot.png")
```

Pass callback to tuner:

```
tuner.tune(
  n_trial=100,
  measure_option=measure_option,
  callbacks=[visualize_callback],
)
```


### With TVM command line

~~Just add a `--visualize` to the `tvmc tune` command.~~

There are three modes available:

- *Live only:* add `--visualize live` to `tvmc tune` command
- *File only:* add `--visualize /pah/to/file.png` to `tvmc tune` command
- *Both:* add `--visualize live,/path/to/file.png` to `tvmc tune` command

## Screenshots

![tvmc_tune_visualize](https://user-images.githubusercontent.com/7712605/185800229-2b84a914-b5df-4afd-86e6-a57747cb81d4.png)



## Additional changes

- Make `si_prefix` in `tvmc/autotuner.py` variable (Prerequisite for a later PR)

## Open Questions
- ~~How to handle the `matplotlib` python dependency? Should I change `python/gen_requirements.py` and if yes which groups (tvmc/dev/?)?~~
- ~~The graph window automatically closes when the tuning is finished. Do we need an option to keep it open until closed?~~ Solved by newly introduced `keep_open` option
- ~~A similar feature for the AutoScheduler would be nice, too. I can look into it if there is interest.~~
- ~~Saving the plot to a file would also be possible. We only need to decide how to expose this to the cmdline (i.e. `--visualize plot.png`)~~

cc @Mousius @gromero

PhilippvK commented:
CC @leandron @giuseros @Mousius @areusch @tqchen

PhilippvK commented:
> @PhilippvK this looks very cool! the only thing missing is to declare the dependency on matplotlib so that the user is not surprised with and ImportError when trying to use it :).
> 

I already had this in mind (see original post) but was unsure on what the best approach is.

> I feel also this should be added to any of the tutorials, because it is too cool for users to miss out.

Yes. I will try to come up with a short section for the tutorials.

What do you think about the other questions?

- The graph window automatically closes when the tuning is finished. Do we need an option to keep it open until closed?
- A similar feature for the AutoScheduler would be nice, too. I can look into it if there is interest.
- Saving the plot to a file would also be possible. We only need to decide how to expose this to the cmdline (i.e. `--visualize-out plot.png`)

areusch reviewed (commented):
@PhilippvK possible to add a unittest for this? 

regarding declaring the dependencies--right now need to do it in two places due to
https://discuss.tvm.apache.org/t/rfc-consolidating-tvm-python-dependencies/8329 ongoing. can you add it to `python/gen_requirements.py` under tvmc extra (we don't have an autotvm one right now) and then if you need to install matplotlib (i think you might in non-gpu container), to docker/install/ubuntu_install_python_package.sh?

PhilippvK commented:
I reworked the PR and updated the comment accordingly. Thank you for your inputs! Please have another look at the changes as I added several improvements:
- AutoScheduler support
- Mention in Tutorial
- Minimal unit test for cmdline parsing of `--visualize` argument
- Support file export

areusch commented on python/tvm/auto_scheduler/task_scheduler.py:
```diff
@@ -659,3 +660,87 @@ def post_tune(self, task_scheduler, task_id):
+        self.out_path = out_path
+        self.init_plot(title)
+
+    def __del__(self):
```
we shouldn't do all this stuff in `__del__`. Tracebacks aren't reported and it can be unpredictable when this is invoked. Can you change this to use `__enter__` and `__exit__` so it can be used with `with` statement, and add an explicit part in the `tvmc` tuning infra where we expect this file written?

areusch commented on python/tvm/auto_scheduler/task_scheduler.py:
```diff
@@ -659,3 +660,87 @@ def post_tune(self, task_scheduler, task_id):
+        self.update_plot()
+
+    def init_plot(self, title):
+        import matplotlib.pyplot as plt
```
can this be moved to the top of the file?

PhilippvK replied:
Since the other callbacks do not need the `matplotlib` Package with would introduce the need to install that package even if that visualize fallback is not used by default. Splitting the  callbacks into separate files would make it at least a bit more concise.

areusch commented on python/tvm/auto_scheduler/task_scheduler.py:
```diff
@@ -659,3 +660,87 @@ def post_tune(self, task_scheduler, task_id):
+        plt.xlabel("Iterations")
+        plt.ylabel(f"{self.si_prefix}FLOPS")
+        if self.live:
+            plt.pause(0.05)
```
can this be 0?

areusch commented on python/tvm/auto_scheduler/task_scheduler.py:
```diff
@@ -659,3 +660,87 @@ def post_tune(self, task_scheduler, task_id):
+                    self.best_flops[i].append(flops)
+            else:
+                flops = None
+        self.update_plot()
```
should we always update or only every N?

areusch commented on python/tvm/autotvm/tuner/callback.py:
```diff
@@ -178,3 +178,82 @@ def _callback(tuner, inputs, results):
+    """
+    import matplotlib.pyplot as plt
+
+    class _Context(object):
```
possible to reuse any of the MS implementation?

PhilippvK replied:
I'll try to refactor the code into a common base class `tvm.contrib.tune_viz.TuneVisualizer` and introduce `AutoTVMVisualizer` as well as `AutoSchedulerVisualizer` inheriting from that.

I haven't touched MS (I guess MetaScheduler) yet, did you mean the AutoScheduler callback?

areusch commented on python/tvm/autotvm/tuner/callback.py:
```diff
@@ -178,3 +178,82 @@ def _callback(tuner, inputs, results):
+            if self.live:
+                plt.pause(0.05)
+
+        def __del__(self):
```
same request here

areusch commented on python/tvm/autotvm/tuner/callback.py:
```diff
@@ -178,3 +178,82 @@ def _callback(tuner, inputs, results):
+    """
+    import matplotlib.pyplot as plt
+
+    class _Context(object):
```
i'm not sure you need a class here to store the vars. they should be captured into the closure below if you reference them from there

areusch commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -139,6 +138,11 @@ def add_tune_parser(subparsers, _, json_params):
         action="store_true",
     )
+    parser.add_argument(
+        "--visualize",
```
this one reads like a boolean option (e.g. action="store_true"). could you better document the behavior?

PhilippvK replied:
@areusch I also had the feeling that this is kinda unintuitive, since I added the option to write the visualizations directly to a file. How about introducing two separate command line flags `—visualize[-live]` (Type: `bool`) and `—visualize-[file|path]` (Type: `str`) to deal with that situation? If you would like that approach, please let me know how the two arguments should be called exactly.

areusch replied:
i think the existing arg is okay, i agree it's a flag with two meanings, but i think if you add it like this, the --help string will be clear:
```
NO_VISUALIZE_OPTION = object()
parser.add_argument("--visualize", nargs="?", type=str, metavar="file", default=NO_VISUALIZE_OPTION, help="Visualize tuning progress. With no parameter specified, display a pyplot window and update the graph as tuning progresses. With a parameter, instead write the graph to the file specified (format determined by the given file extension).")
```

(I think the "format determined by the given file extension" is correct, but maybe double check that)

Alternatively open to the two options, especially if you think users want both. you could also consider something like `--plot=live,foo.png`, which might provide a nice set of flexibility without users needing to know two options and how they interact.

PhilippvK replied:
@areusch 
> `--plot=live,foo.png`

This should be what is implemented right now just with a different name. The main disadvantage here is that the custom format/parsing of the argument (i.e. `live, path.png`, `path.png,live`,...) can be confusing.

> `parser.add_argument("--visualize", nargs="?",   ...`

Your proposed solution which accepts either one or no args at all, looks nice at the first glance but the `nargs="?"` might lead to some unexpected behavior if a model path is confused with the argument to the `--visualize` option by argparse. In addition, as you already suggested this only allows:
- No visualizations
- Live only
- File only

That's why I would prefer the two separate options, which also allow a 4 combinations:
- No visualizations
- Live only
- File only
- Both

areusch commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -148,20 +152,17 @@ def add_tune_parser(subparsers, _, json_params):
         type=int,
-        help="the size of cache line in bytes. "
-        "If not specified, it will be autoset for the current machine.",
+        help="the size of cache line in bytes. " "If not specified, it will be autoset for the current machine.",
```
can you join the strings here and below?

areusch commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -228,6 +226,30 @@ def add_tune_parser(subparsers, _, json_params):
+        else:
+            assert file is None, "Only a single path can be passed to the --visualize argument"
+            file = pathlib.Path(item)
+    mode = "none"
```
can you create an Enum class to document mode?

areusch commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -596,6 +636,8 @@ def schedule_tasks(
     tuning_options: auto_scheduler.TuningOptions,
     prior_records: Optional[str] = None,
     log_estimated_latency: bool = False,
+    visualize_mode: str = "none",
```
can you update this type to reflect the Enum class?

areusch commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -638,6 +685,9 @@ def tune_tasks(
     early_stopping: Optional[int] = None,
     tuning_records: Optional[str] = None,
+    si_prefix: str = "G",
+    visualize_mode: str = "none",
```
here as well

areusch commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -691,12 +747,25 @@ def tune_tasks(
+        ]
+
+        if visualize_mode != "none":
+            assert visualize_mode in ["both", "live", "none"]
```
update this to `assert visualize_mode in VisualizeModes`

areusch commented on tests/python/driver/tvmc/test_autotuner.py:
```diff
@@ -182,3 +182,18 @@ def test_tune_rpc_tracker_parsing(mock_load_model, mock_tune_model, mock_auto_sc
     assert "10.0.0.1" == kwargs["hostname"]
     assert "port" in kwargs
     assert 9999 == kwargs["port"]
+
```
possible to add a test using e.g. pyplot svg backend and a golden svg file?

PhilippvK replied:
I will look into this.

areusch reviewed (changes requested):
thanks @PhilippvK and apologies I've been slow to review. This is really neat!

Can you address the comments here particularly the ones around the use of `__del__`?

PhilippvK commented:
I will followup with a new branch next year!
